src/runtime/runtime.go

```diff
@@ -88,13 +88,45 @@ func LockOSThread() {
 func UnlockOSThread() {
 }
 
+// Mark KeepAlive as noinline so that it is easily detectable as an intrinsic.
+//
+//go:noinline
+
+// KeepAlive marks its argument as currently reachable.
+// This ensures that the object is not freed, and its finalizer is not run,
+// before the point in the program where KeepAlive is called.
+//
+// A very simplified example showing where KeepAlive is required:
+//
+//	type File struct { d int }
+//	d, err := syscall.Open("/file/path", syscall.O_RDONLY, 0)
+//	// ... do something if err != nil ...
+//	p := &File{d}
+//	runtime.SetFinalizer(p, func(p *File) { syscall.Close(p.d) })
+//	var buf [10]byte
+//	n, err := syscall.Read(p.d, buf[:])
+//	// Ensure p is not finalized until Read returns.
+//	runtime.KeepAlive(p)
+//	// No more uses of p after this point.
+//
+// Without the KeepAlive call, the finalizer could run at the start of
+// syscall.Read, closing the file descriptor before syscall.Read makes
+// the actual system call.
+//
+// Note: KeepAlive should only be used to prevent finalizers from
+// running prematurely. In particular, when used with unsafe.Pointer,
+// the rules for valid uses of unsafe.Pointer still apply.
 func KeepAlive(x interface{}) {
-	// Unimplemented. Only required with SetFinalizer().
+	if alwaysFalse {
+		println(x)
+	}
 }
 
-func SetFinalizer(obj interface{}, finalizer interface{}) {
-	// Unimplemented.
-}
+// alwaysFalse is a boolean value that is always false.
+// The compiler cannot see that alwaysFalse is always false,
+// so it emits the test and keeps the call, giving the desired
+// escape analysis result. The test is cheaper than the call.
+var alwaysFalse bool
```

Comment:
In Go, the variable is named `cgoAlwaysFalse` but I think this was just to reuse the same variable after the fact.

https://github.com/golang/go/blob/master/src/runtime/mfinal.go#L515

Reply:
I'm not convinced the new LTO logic will think `alwaysFalse` is potentially not always `false`.  I think we need something else here that we can ensure at the compiler level.

Reply:
What do you think about just noinline being enough? If it seems so I could remove this stuff to prevent confusion.

Reply:
Went ahead and removed down to just adding go:noinline, verifying it actually does keepalive is probably quite tricky but in the meantime it can be attempted along with finalizers in custom GC.